### DISCUSSION TITLE
Desktop client with self-hosted server

### DISCUSSION THREAD

**Discussion by brianclemens** [outline/outline#5512] [General] [answered]
Is it possible to use the desktop client with a self-hosted server? Clicking "settings" doesn't actually open anything, I was hoping it might open a panel where I could plug in my custom URL or something.

**Comment by arnolem** [comment]
Hello, 
I have the same need but for your information, you can install the PWA app with self-hosted :
https://www.getoutline.com/changelog/progressive-web-app

**Comment by Djiit** [comment]
Hey team, I would also need this feature. Is the desktop app source code open source? Happy to contribute this.

**Comment by M4rlus** [comment]
@brianclemens @arnolem @Djiit 
Go into chrome, open up your outline, click on the three dots at the right top of chromes addon bar and click install outline.
It works like a real desktop app.

**Reply by brianclemens** [reply]
@M4rlus I appreciate the suggestion but a PWA isn't what I'm looking for, thank you.

**Reply by Djiit** [reply]
Thanks. Actually I don't see the option anywhere (I do see it for other services). Do I need a specific setup? I'm using an up to date self hosted instance

**Comment by wouter-deen** [comment]
Would love to see this. Although the PWA works fine, the actual desktop apps are more polished (although it looks like they're still Electron-based or similar). Don't know if this is relevant to this topic, but it would also make it possible for a (hopefully) future mobile app to be compatible with a self-hosted instance. I personally just use Outline for myself as a Notion replacement, as it is simply the best alternative out there.

**Comment by deviant** [comment]
Until this is implemented, self-hosters using macOS can work around the missing functionality by doing the following:

1. Download and open the app once, to make it pass Gatekeeper, then quit it again.
2. Run `rm -r ~/Library/Application\ Support/Outline`, to clear the app cache.
3. Install https://github.com/electron/asar (if using node/npm, run `npm install -g @electron/asar`)
4. Run `asar extract /Applications/Outline.app/Contents/Resources/app.asar outline-app`
5. Run `sed -i '' 's/app.getoutline.com/<your Outline domain>/' outline-app/build/env.js`
6. Run `asar pack outline-app /Applications/Outline.app/Contents/Resources/app.asar`
7. Open the app. You may not be able to complete a sign-in flow— in that case:
8. Open Outline in your browser and authenticate. Extract the `accessToken` cookie by some means (probably DevTools).
9. Back in the app, open DevTools using <kbd>Option-Command-I</kbd> and open the Console.
10. Enter `document.cookie = "accessToken=<your access token>"`
11. Close DevTools and reload the page with <kbd>Command-R</kbd>.

This should leave you logged into your personal Outline instance, in the app. I've not been using this for long, so I'm not sure how long it's stable for, but some of these steps will obviously need to be re-done when updating the app or when your session expires for whatever reason.

Instructions for Windows should be similar, albeit with different paths. Perhaps someone else would like to write an adaptation for it and share them with us.

**Comment by matthiscock** [comment]
Makes Self Hosters feel like sloppy seconds.......

**Comment by Dorifor** [comment]
any news on this ? I don't want my apps to just be another tab in my browser.

**Reply by tmeuze** [reply]
Agreed, fair - but PWAs are generally designed to be distinguishable from regular tabs in a browser. Safari and Chrome (last I checked) support this type of independent instance.

**Comment by tmeuze** [comment]
I hate to be that guy, but it's appropriate to think that the Desktop/Mobile apps serve as an incentive to choose the hosted variant of Outline. Have to make money somehow. 

Could be a nice sponsorship feature if feasible.

**Reply by mrvnklm** [reply]
I agree - this is not the right way to make people feel they have to get the hosted version. If I want my own data on my server, I should have limited functionality but not limited accessibility.

**Comment by kaivol** [comment]
I'm a bit confused, did the developers ever state whether the desktop client will be coming for self-hosted servers or not?

**Reply by tmeuze** [reply]
No indication either way, afaik. Just not implemented currently.

**Comment by psychedelicu** [comment]
Any news? I want to use windows app, but i can't see how to login on the self-hosted outline.

**Comment by jfbloom22** [comment]
I was so excited to self host Outline and run a desktop app.  So bummed this is not supported!

I did some poking around and updated my system hosts file to redirect from getoutline.com to my host, but I can't login.  The Google auth fails.  For Outline to support this, I believe we need to update the Google auth Client ID in the desktop app.  

Will use the Safari "Add to Dock" feature to turn it into a desktop app.  Am I missing any features they built into the official desktop app?

**Reply by matthiscock** [reply]
You should not update your hosts files.
You need to install on a server which is linked to a domain

E.g. we have ours setup on a 'server' (in this case its on an unraid install) and then using nginx reverse proxy we link that up to a domain
so we have outline.mydomain.com - which points towards our private ip (from our ISP) and then routes to the internal port and ip within our network

Hope that helps

**Reply by scottrbaxter** [reply]
@matthiscock where a domain is useful and recommended for production hardening, this isn't a hard & fast requirement. it's not necessary in a dev environment, or on a dedicated system without direct internet access. using a hosts file is a very useful feature for dev and testing locally, it's also very easy to spin up something like Outline to use on a standalone machine, so long as you don't _need_  to share with anyone.

**Reply by jfbloom22** [reply]
My comment might be a little confusing. I have Outline deployed to a server. I wanted to use the Outline desktop app, but I don't see an option to point it at my server domain. I inspected the traffic of the desktop app and updated my local hosts file so that Outline connected to my server domain.  It worked, but Auth fails. 

I expect I could get it working if I did my own auth. Saw someone else doing custom auth. But I don't want to go through that much trouble.

**Comment by burritosoftware** [comment]
I am working on a mod you can patch into Outline Desktop that allows for custom instance URLs to be changed on the fly. A preview is below and I’ll update this when I have documentation setup. It consists of two parts: a patcher to patch the client, and a redirector to pass your login token from Outline web to desktop.

https://github.com/user-attachments/assets/92372e01-5160-4e7d-ae3e-706afd62f3a6

**Reply by psychedelicu** [reply]
This is actually amazing! The community will thank you!

**Reply by matthiscock** [reply]
How close is a realease for this please?
Even if CLI Based

Thanks for your hard work!

**Reply by burritosoftware** [reply]
I’m currently attending an event over the weekend but a CLI release should be up around next week. I have a couple of things to fix up with it: notably I want to disable Outline’s Sentry analytics when using this mod as I don’t want them to receive potential errors and bug reports related to it! 😅

Also, I’m going to be making sure the patcher works for macOS as well. The mod should work fine but just the installation needs to be ported to work on that platform. Windows is already ready to go patcher-side, just need to fix up the mod itself.

**Reply by zkvvoob** [reply]
I am quite looking forward to seeting your patcher for macOS. If you need any testing let me know – I'd be happy to help in any way I can (I also know my way around Python).

On a side note, I suppose it would be too much to hope that the pached desktop client would be able to use the self-hosted instance's authentication method? I'm asking because mine use Authelia and not Google, Microsoft or Slack.

**Reply by c581698** [reply]
@burritosoftware If it's not too much to ask are you still working on it? i cound't find the [outline-desktop-patcher](https://github.com/burritosoftware/outline-desktop-patcher) anymore.

**Reply by burritosoftware** [reply]
Hi, yeah! Sorry, I got swamped with a bunch of stuff, but I can release it soon (maybe this week). Just been extremely busy with other things

**Reply by Novapixel1010** [reply]
This off topic. BUT how did you imbed the video? 😆 Does it just have to be .mov and not .mp4

**Comment by burritosoftware** [comment]
Hi everyone, my mod for Outline Desktop working with custom instances is now out in a pre-release state.

Will still continue to work on it, as it is missing a couple things that I still need to work on, and the documentation is kinda sprawly, but I'll try to get everything central in this post so you can try to get it working for yourself! (And if anyone wants to contribute, check out the three repositories for the project below. This is heavily inspired by Vencord but in a way simpler fashion.)

---

# Repositories
[patcher](https://github.com/burritosoftware/outline-desktop-patcher): Patches an Outline installation by installing the mod.
[mod](https://github.com/burritosoftware/outline-desktop-mod): The actual mod for Outline, with a plugin system (comes with [`custom-instances`](https://github.com/burritosoftware/outline-desktop-mod/blob/master/corePlugins/custom-instances.js) and [`disable-sentry`](https://github.com/burritosoftware/outline-desktop-mod/blob/master/corePlugins/disable-sentry.js) plugins)
[redirector](https://github.com/burritosoftware/outline-desktop-redirector): Fixes desktop logins for self-hosted Outline instances

# Known Limitations
- Only on Windows 10/11 for now (still working on macOS support)
- The binary may not be able to patch system installations. Either run the patcher from source, or install a per-user installation

# Usage
Download the .exe from [here](https://github.com/burritosoftware/outline-desktop-patcher/releases/tag/0.1.0) and run it. It should detect all your Outline installations. Then you can patch, and after you patch, you can set a custom instance URL. **Once you're done, you need to follow the login setup below, otherwise you'll be stuck trying to login!**

If you want to run from source, [follow the instructions in the repo](https://github.com/burritosoftware/outline-desktop-patcher).

# Login Setup
Unfortunately, patching isn't enough to get your custom instance working inside the app, as self-hosted instances don't provide the callback to pass your account login token back to the desktop app automatically.

You can fix this in two ways: automatically using Cloudflare Workers, or manually by grabbing your token from your browser.  
Automatic is recommended so that you don't have to manually transfer your token every time you're logged out due to inactivity.

## Automatic with Cloudflare
To fix this, you want to make sure your Outline instance is behind Cloudflare, then [deploy my custom worker](https://github.com/burritosoftware/outline-desktop-redirector). Simply click the Deploy to Cloudflare button and follow all prompts to deploy.

Once you do, go to the Cloudflare Dashboard, access the Worker and go to its settings. Add a route to the Worker that points to `<your-outline-instance>/desktop-login` in **"Fail closed" mode**.

Then, you need to add any of the routes below necessary in the same manner in **"Fail open" mode** this time but depending on what login providers you use.
```
/auth/oidc*
/auth/email*
/auth/azure*
/auth/slack*
/auth/google*
/auth/discord*
```

An example for OIDC-only looks like this: 
![image](https://github.com/user-attachments/assets/c75db683-ad62-45a3-a62f-20becefa2d29)

Lastly, create a document in Outline that has a link to `<your-outline-instance>/desktop-login`. This will help users manually initiate a login when they are logged out on browser (the Worker requires that you are logged in on browser to retrieve your token and pass back, if not it will redirect to a standard login flow and unfortunately can't recover after someone has logged in with your login provider).  
Example:
![image](https://github.com/user-attachments/assets/7270d708-036f-4688-b2d0-28d6cfbdd5eb)

You're done! Open the Outline Desktop app and you should be redirected back to it after your browser passes back your token. If you are forced to re login in browser, you will need to use the manual link we created in the document earlier to login.

## Manual with Browser
Login to your Outline instance in your browser. Then open DevTools, and go to Application > Storage (or on Firefox, Storage).
Under the dropdown for Cookies, click your instance URL, then find a key named `accessToken`. Copy the value of that.

Then go to `<your-outline-instance>/desktop-redirect?token=<accessToken>`, and accept the prompt to open the Outline app.

---

Hopefully you got it up and running! Feel free to comment here if you have any issues. I've been swamped recently with a bunch of life stuff, so if you can look at the source and understand a bit at what's going on, I accept contributions and PRs! Thanks :3

**Reply by burritosoftware** [reply]
@brianclemens If you wouldn't mind marking this as answered, would be appreciated! I don't think the Outline team is going to officially make their desktop app work on self-hosted instances unfortunately :(

**Reply by lukzmu** [reply]
The fact that this discussion is already 2 years old and there's not a single official answer from the Outline team. Well this isn't looking promising 😞.

**Reply by brianclemens** [reply]
Yeah, I ditched Outline a long time ago. However, I won't mark it as answered, as a mod is not what I (and likely others here) are looking for. It's neat, just not the intent of the issue. I was looking for support in the official client. I'll just leave this open and unsubscribe.

**Reply by psychedelicu** [reply]
Yeah, I switched back to Notion until they will implement the Windows client.

**Reply by cdfatpitch** [reply]
@burritosoftware still working on the MacOS patch? asking for a friend

**Reply by silverjerk** [reply]
@brianclemens what did you end up moving to? I've run through most of the self-hosted solutions that work with my MCP servers, as I rely heavily on this to generate and update my homelab's documentation. That's included Docmost, Outline, Affine, AppFlowy, and numerous others. If it weren't for Affine's inability to actually push raw markdown via their API, I wouldn't have circled back to try and get Outline working again with the local MacOS client. 

TL;DR: I haven't come across a self-hosted documentation solution where the desktop client works with the self-hosted server without a non-starter somewhere in the installation or use of the app.

**Reply by tmeuze** [reply]
Just wanted to make notice that this has since been resolved (a few months ago, just found out).

**Comment by mmmeff** [comment]
Hey @tommoor I built y'alls MCP server so you should add support for this for my team 🤣 . Joking of course, but yeah this would be killer for those of us hosting enterprise instances of Outline.

**Comment by Pezhvak** [comment]
Soo, no support in 2026?

**Reply by oregapam** [reply]
+1

**Reply by DavidRadobenko** [reply]
+1

**Reply by tmeuze** [reply]
Available since 23rd June 2026 - see below.

**Comment by tmeuze** [comment, marked as answer]
Hey all, crazy that I just realized, but - 

Since 23rd June 2026, [Outline Desktop app supports self-hosted instances](https://www.getoutline.com/changelog/desktop-improvements):
> Installation switcher – a new switcher on the login screen lets you move between different Outline installations, making the desktop app compatible with on-premise deployments.

Thank you so much to @tommoor and the Outline team! Think this is case closed.

**Reply by oregapam** [reply]
I can confirm it works like a charm 👍

**Reply by tmeuze** [reply]
Been testing it out. Unfortunately if trying to add a second instance, the server switcher menu just prompts you to input a hosted subdomain (`*.getoutline.com`). Hopefully that can be ironed out. 

<img width="1212" height="586" alt="csX_Outline - 2026-09-18 at 12 50 00@2x" src="https://github.com/user-attachments/assets/b86bb7c7-f448-4c87-a920-20593c6115cc" />
